### ISSUE TITLE
ttnn.div(a, inf) returns NaN instead of 0 in float32 — the residual refinement poisons a correct result

### ISSUE THREAD

**Issue by thwaitsusa0316** [tenstorrent/tt-metal#52112]
## Summary

With fp32 DEST accumulation enabled, `ttnn.div(a, inf)` for any finite `a` returns `NaN`.
IEEE-754 requires `0`. The multiply that produces the quotient gets the right answer; the
Markstein residual refinement that runs afterwards destroys it.

The registered golden for this op is `torch.div` (`ttnn/ttnn/operations/binary.py#L321-L327`),
and `torch.div(torch.tensor(1.0), torch.tensor(float("inf")))` is `0.0`, so the expected value
is not in dispute.

## Where

`calculate_sfpu_binary_div`, in the `is_fp32_dest_acc_en` branch:

- blackhole: `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h`
- wormhole_b0: `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h`

```cpp
sfpi::vFloat r = sfpu_reciprocal_iter<2>(in1);
sfpi::vFloat result = in0 * r;
if constexpr (is_fp32_dest_acc_en) {
    v_if(sfpi::exexp(result, sfpi::ExponentMode::Biased) != 255) {
        sfpi::vFloat e = in0 - result * in1;
        result = result + e * r;
    }
    v_endif;
}
```

## Trace

For finite `a` and `b = +inf`:

| step | value |
|---|---|
| `r = sfpu_reciprocal_iter<2>(inf)` | `+0` — documented in `ckernel_sfpu_recip.h` |
| `result = a * 0` | `0` — **this is already the correct answer** |
| guard `exexp(result) != 255` | biased exponent of `0` is `0`, so the guard is **true** and refinement runs |
| `e = a - result*b = a - (0 * inf)` | `0 * inf` is `NaN`, so `e = NaN` |
| `result = 0 + NaN * 0` | **`NaN`** |

The existing guard only skips refinement when the *quotient* is already non-finite — the
`a/b → inf` overflow case. It does not consider a non-finite *denominator*, which is the case
where the residual `result * den` is an invalid `0 * inf` operation.

The trailing special-case block handles `in1 == 0` (`x/0 → ±inf`, `0/0 → NaN`). There is no
corresponding handling for `in1 == ±inf`.

## Scope

- Affects the `is_fp32_dest_acc_en` path of `calculate_sfpu_binary_div` only.
- The plain `BinaryOp::DIV` case in `calculate_sfpu_binary` has no refinement, so `a/inf` is
  `a * 0 = 0` there and appears correct.
- Quasar has no Markstein block in its div path and appears unaffected.
- `div_no_nan` shares the helper and should be checked alongside.

## Why the tests miss it

`test_binary_fp32.py::test_div_fp32` includes an `inf / inf` pair, which correctly yields `NaN`
and passes. It does not test **finite / inf**, which is the failing combination. The sweeps draw
both operands from `[-100, 100]` and never generate infinities at all.

## Suggested fix

Extend the refinement guard to also skip when the denominator is non-finite, e.g.
`v_if(exexp(result, Biased) != 255 && exexp(in1, Biased) != 255)`, or equivalently skip when
`r == 0`. Either keeps `result = a * 0 = 0` and leaves every finite case untouched.

## Caveat on verification

I do not have hardware, so this is derived from the source plus a float32 model of the code path
rather than measured on device. It assumes the SFPU multiply follows IEEE for `0 * inf → NaN`.
That assumption is consistent with this file's own comments, which already reason about
`inf + (-inf) = NaN` in the guard directly above. If the SFPU instead defines `0 * inf` as `0`,
this does not reproduce and the issue can be closed — a one-line check on device would settle it
either way.

## Why it matters

Like #52094, this is not precision loss: it is a wrong, non-finite answer where the correct
answer is an ordinary `0`. Unlike a silent `0`, a `NaN` does propagate loudly — but it appears in
exactly the code that is trying to be *more* accurate, and it turns a case the hardware already
got right into a failure. Dividing by an accumulated magnitude that has overflowed to `inf` is a
realistic way to hit it.

**Comment by thwaitsusa0316** [comment]
My apologies...

I need to disclose something up front.  This issue was submitted through tooling rather than posted by me by hand, which I now see conflicts with the AI Tool Restrictions in CONTRIBUTING.md — specifically the line prohibiting AI agents from posting directly to GitHub issues, and the requirement that AI-assisted content be posted manually by the human contributor.  That was my mistake and I'm flagging it rather than leaving it to be noticed.

To be clear about what this is and isn't: it's a bug report, not a claim on a bounty or a request for assignment.  However my posting mechanism was wrong either way, so i'm withdrawing it and redo'ing it properly.  I stand behind and take full responsibility

On the substance, I read the kernel source, traced the refinement path, and checked the math myself.  I understand the change I'm describing and can defend it. I leverage AI assistance for analysis and drafting, which as I read CONTRIBUTING.md is permitted for a contributor's own development work — the error was in how it reached GitHub.

Re-filing the same report manually now. Happy to answer any questions about the policy or to take a different course if the maintainers would prefer — @jberkowitzTT.  Once again, apologies for any inconvenience. steve

**Comment by jberkowitzTT** [comment]
Hi @thwaitsusa0316 - thank you so much for your thoughtfulness on this, and we appreciate your sincerity. If this is a legitimate bug you believe to have found, the team will review the submission. Let me if you at any point interested in working on this or another idea you have as a bounty. Before submitting a bounty issue, please reach out to bounties@tenstorrent.com for review.

Thanks!

**Comment by thwaitsusa0316** [comment]
just to confirm, you want me to email the bounty issue to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com) before submitting?  I just posted 52114 (before reading this comment - sorry).  

(I am working on the email) 
thanks

**Comment by jberkowitzTT** [comment]
Not a problem! If you're specifically looking to create a bounty issue, send an email first. Otherwise (like if you're just looking to file a bug or work on an issue but not as a bounty), there's no need to send an email

**Comment by thwaitsusa0316** [comment]
Email has been sent!!!
